mesonbuild/interpreter.py

```diff
@@ -2255,8 +2255,21 @@ def func_configure_file(self, node, args, kwargs):
 
     @stringArgs
     def func_include_directories(self, node, args, kwargs):
-        absbase = os.path.join(self.environment.get_source_dir(), self.subdir)
+        src_root = self.environment.get_source_dir()
+        absbase = os.path.join(src_root, self.subdir)
         for a in args:
+            if a.startswith(src_root):
+                raise InvalidArguments('''Tried to form an absolute path to a source dir. You should not do that but use
+relative paths instead.
+
+To get include path to the current dir do this:
+
+incdir = include_directories('.')
+
+After this incdir will contain both the current source dir as well as the
+corresponding build dir. It can then be used in any subdirectory and
+Meson will take care of all the busywork to make paths work.
```

Comment:
Should also mention here that you can use any relative path to any subdirectory otherwise people will think they have to create `meson.build` files in directories that only contain headers just to get that value set.